compose/cli/log_printer.py

```diff
@@ -36,10 +37,18 @@ def no_color(text):
         else:
             color_funcs = cycle(colors.rainbow())
 
-        for color_func, container in zip(color_funcs, self.containers):
-            generator_func = get_log_generator(container)
-            prefix = color_func(build_log_prefix(container, prefix_width))
-            yield generator_func(container, prefix, color_func)
+        if self.log_streams:
+            # we're logging from the `up` command and already have a list of
+            # threads attached to containers
+            for color_func, container, stream in zip(color_funcs, self.containers, self.log_streams):
+                generator_func = build_log_generator
+                prefix = color_func(build_log_prefix(container, prefix_width))
+                yield generator_func(container, prefix, color_func, stream)
+        else:
+            for color_func, container in zip(color_funcs, self.containers):
+                generator_func = get_log_generator(container)
+                prefix = color_func(build_log_prefix(container, prefix_width))
+                yield generator_func(container, prefix, color_func)
```

Comment:
I think a lot more of this code can be shared between the branches by making a small change to `get_log_generator()`. I was hoping to have a chance today to prototype that, but I've run out of time.

I think we can get it down to a single line in each if/else branch inside the for loop